libstorage/RocksDBStorage.cpp

```diff
@@ -0,0 +1,220 @@
+            for (auto it = values.begin(); it != values.end(); ++it)
+            {
+                Entry::Ptr entry = std::make_shared<Entry>();
+
+                for (auto valueIt = it->begin(); valueIt != it->end(); ++valueIt)
+                {
+                    entry->setField(valueIt.key().asString(), valueIt->asString());
+                }
+
+                if (entry->getStatus() == Entry::Status::NORMAL && condition->process(entry))
+                {
+                    entry->setDirty(false);
+                    entries->addEntry(entry);
+                }
+            }
+        }
+
+        return entries;
+    }
+    catch (std::exception& e)
+    {
+        STORAGE_LEVELDB_LOG(ERROR) << LOG_DESC("Query rocksdb exception")
+                                   << LOG_KV("msg", boost::diagnostic_information(e));
+
+        BOOST_THROW_EXCEPTION(e);
+    }
+
+    return Entries::Ptr();
+}
+
+size_t RocksDBStorage::commit(h256 hash, int64_t num, const std::vector<TableData::Ptr>& datas)
+{
+    try
+    {
+        auto hex = hash.hex();
+
+        WriteBatch batch;
+        for (size_t i = 0; i < datas.size(); ++i)
+        {
+            std::shared_ptr<std::map<std::string, Json::Value> > key2value =
+                std::make_shared<std::map<std::string, Json::Value> >();
+
+            auto tableInfo = datas[i]->info;
+
+            processEntries(hash, num, key2value, tableInfo, datas[i]->dirtyEntries);
+            processEntries(hash, num, key2value, tableInfo, datas[i]->newEntries);
+
+            for (auto it : *key2value)
+            {
+                std::string entryKey = tableInfo->name + "_" + it.first;
+                std::stringstream ssOut;
+                ssOut << it.second;
+
+                batch.Put(Slice(entryKey), Slice(ssOut.str()));
+            }
+        }
+
+        m_db->Write(WriteOptions(), &batch);
+        return datas.size();
+    }
+    catch (std::exception& e)
+    {
+        STORAGE_LEVELDB_LOG(ERROR) << LOG_DESC("Commit rocksdb exception")
+                                   << LOG_KV("msg", boost::diagnostic_information(e));
+        BOOST_THROW_EXCEPTION(e);
+    }
+
+    return 0;
+}
+
+bool RocksDBStorage::onlyDirty()
+{
+    return false;
+}
+
+void RocksDBStorage::setDB(std::shared_ptr<rocksdb::DB> db)
+{
+    m_db = db;
+}
+
+void RocksDBStorage::processEntries(h256 hash, int64_t num,
+    std::shared_ptr<std::map<std::string, Json::Value> > key2value, TableInfo::Ptr tableInfo,
+    Entries::Ptr entries)
+{
+    for (size_t j = 0; j < entries->size(); ++j)
+    {
+        auto entry = entries->get(j);
+        auto key = entry->getField(tableInfo->key);
+
+        auto it = key2value->find(key);
+        if (it == key2value->end())
+        {
+            std::string entryKey = tableInfo->name;
+            entryKey.append("_").append(key);
+
+            std::string value;
+            auto s = m_db->Get(ReadOptions(), Slice(entryKey), &value);
```

Comment:
Also need to append all the history value of a given key, mark this, and to see if we can optimize it in the future version.